hydroshare_jupyter_sync/lib/filesystem/fs_resource_map.py

```diff
@@ -0,0 +1,199 @@
+
+    @abstractmethod
+    def update_resource(self) -> None:
+        """Update entire resource map. md5hash's updated, non-existent files removed, and new files inserted."""
+
+
+class IEntityFSResourceMap(ABC):
+    """Class representing the relationship between a file (no directory) path to the file's MD5 Hash."""
+
+    @abstractmethod
+    def add_file(self, relative_resource_file: Union[Path, str]) -> None:
+        """Add file to FSResource map."""
+
+    @abstractmethod
+    def delete_file(self, relative_resource_file: Union[Path, str]) -> None:
+        """Remove file from FSResource map."""
+
+    @abstractmethod
+    def update_file(self, relative_resource_file: Union[Path, str]) -> None:
+        """Update file in FSResource map."""
+
+
+# semi-concrete implementations
+
+
+class FSResourceMap(UserDict, IFSResourceMap):
+    @property
+    def files(self) -> List[Path]:
+        """Return list of files in resource."""
+        return list(self.data.keys())
+
+
+# concrete implementations
+
+
+class LocalFSResourceMap(FSResourceMap, IEntityFSResourceMap):
+    """Concrete class representing the relationship between a local file (not directory) path to the file's MD5 Hash."""
+
+    def __init__(self, resource_path: Union[Path, str]) -> Dict[Path, MD5Hash]:
+        super().__init__()
+        self.resource_path = Path(resource_path).expanduser().resolve()
+        self.resource_id = resource_path.name
+
+    @classmethod
+    def from_resource_path(
+        cls, resource_path: Union[Path, str]
+    ) -> "LocalFSResourceMap":
+        # create class instance
+        fsresource_map = cls(resource_path)
+
+        fsresource_map.update_resource()
+        return fsresource_map
+
+    def add_file(self, relative_resource_file: Union[Path, str]) -> None:
+        if self._valid_resource_file(relative_resource_file) and not self._is_member(
+            relative_resource_file
+        ):
+            self._insert(relative_resource_file)
+
+    def update_file(self, relative_resource_file: Union[Path, str]) -> None:
+        if self._is_member(relative_resource_file):
+            self._insert(relative_resource_file)
+
+    def delete_file(self, relative_resource_file: Union[Path, str]) -> None:
+        if self._is_member(relative_resource_file):
+            relative_resource_file = self._as_child_of_base_directory(
+                relative_resource_file
+            )
+            del self.data[relative_resource_file]
+
+    def update_resource(self) -> None:
+        # clear data dictionary
+        self.data = dict()
+        for resource_file in self.data_path.glob("**/*"):
+            # insert relative file path to data_path and md5 digest
+            self._insert(resource_file)
+
+    @property
+    def base_directory(self) -> Path:
+        """Return assumed path to base directory (i.e. `/some/path/{resource_id}/{resource_id}`).
```

Comment:
There are two folders with resource_id?  I assume this has something to do with "open with" on hydroshare with the automatic bag download/unzip.  This folder structure is not guaranteed if a user downloads/unzips manually.

Reply:
That's a great point. Maybe this is not the best assumption.

Reply:
Ill create an issue based on this and circle back to it in the future.